src/main/java/io/nats/client/impl/NatsConnection.java

```diff
@@ -1524,30 +1512,72 @@ public long getMaxPayload() {
     }
 
     public Collection<String> getServers() {
-        NatsServerInfo info = this.serverInfo.get();
-        HashSet<String> check = new HashSet<>();
-        ArrayList<String> servers = new ArrayList<>();
-
-        options.getServers().stream().forEach(x -> {
-            String uri = x.toString();
-            if (!check.contains(uri)) {
-                servers.add(uri);
-                check.add(uri);
-            }
-        });
+        return getServersList();
+    }
+
+    private List<String> getServersList() {
+        return addDiscoveredServers( addConfiguredServers(new ArrayList<>()) );
+    }
+
+    private List<String> buildServerList() {
+        if (options.isNoRandomize()) {
+            return getServersList();
+        }
+
+        // prioritize discovered over configured
+        List<String> servers = shuffleWithoutCurrent( addDiscoveredServers(new ArrayList<>()) );
+
+        // configured get shuffled
+        servers.addAll( shuffleWithoutCurrent( addConfiguredServers(new ArrayList<>()) ) );
+
+        // because we still want it in the list
+        if (currentServer != null) {
+            servers.add(currentServer);
+        }
+

```

Comment:
get the discovered servers and shuffle them (without the current). separately get the configured (in options) and shuffle them (also w/o current) then add the configured and then the current to the discovered list and return